src/proposal.mligo

```diff
@@ -341,6 +365,65 @@ let delete_proposal
     Set.remove (start_date, proposal_key) store.proposal_key_list_sort_by_date
   }
 
+[@inline]
+let fraction_to_quorum_fraction(n, d : nat * nat): unsigned_quorum_fraction
+  = { numerator = (n * quorum_denominator) / d }
+
+[@inline]
+let to_signed(n : unsigned_quorum_fraction): quorum_fraction
+  = { numerator = int(n.numerator) }
+
+[@inline]
+let to_unsigned(n : quorum_fraction): unsigned_quorum_fraction
+  = { numerator = match is_nat(n.numerator) with
+              | Some n -> n
+              | None -> (failwith("BAD_UNSIGNED_CONVERSION"):nat)
+    }
+
+let update_quorum(store, config : storage * config): storage =
+  let current_period = get_current_period_num(store.start_time, config.voting_period) in
+  if store.quorum_threshold_at_stage.last_updated_period = current_period
+    then store // Quorum has been updated in this period, so no change is required.
+    else
+      match is_nat(current_period - store.period_staked.last_updated_period) with
+        | Some elapsed -> if elapsed >= 2n
+            then
+              let previous_staked = store.period_staked.staked in
+              let previous_participation = to_signed(fraction_to_quorum_fraction(previous_staked, config.governance_total_supply)) in
+              let old_quorum = to_signed(store.quorum_threshold_at_stage.quorum_threshold) in
+              let quorum_change = to_signed(config.quorum_change) in
+              let possible_new_quorum =
+                // old_quorum + (previous_participation - old_quorum) * quorum_change
+                fadd(old_quorum, fmul(quorum_change, fsub(previous_participation, old_quorum))) in
+                //old_quorum + fmul(quorum_change, previous_participation - old_quorum) in
+              let one_plus_max_change_percent = to_signed({ numerator = config.max_quorum_change.numerator + quorum_denominator }) in
+              let min_new_quorum =
+                fdiv(old_quorum, one_plus_max_change_percent) in
+              let max_new_quorum =
+                fmul(old_quorum, one_plus_max_change_percent) in
+
+              let config_min_qt = to_signed(config.min_quorum_threshold) in
+              let config_max_qt = to_signed(config.max_quorum_threshold) in
+              let new_quorum = bound_qt(bound_qt(possible_new_quorum, min_new_quorum, max_new_quorum), config_min_qt, config_max_qt)
+              in { store with quorum_threshold_at_stage =
+                   { quorum_threshold = to_unsigned(new_quorum)
+                   ; last_updated_period = current_period
+                   }
+                 }
```

Comment:
As an argument in favor of allowing both `voting` and `proposing` stage numbers, here what happens if nobody calls this function for the whole length of a `proposing` stage?

AFAIU in the following period the `elapsed >= 2` check will pass and here we'll store a `voting` stage number.

Reply:
>AFAIU in the following period the elapsed >= 2 check will pass and here we'll store a voting stage number.

Basically since we call update_quorum in `propose` operation, I was counting on it not happening in a voting period. But now I have changed the whole thing to use cycles anyway.

Reply:
Oh I see, that's a good point about `vote`, we indeed don't need to check for an update there as it cannot happen in a different cycle than the last `propose` call.